src/dsp_tools/analyse_xml_data/construct_and_analyze_graph.py

```diff
@@ -87,62 +87,66 @@ def _extract_ids_from_one_text_value(text: etree._Element) -> set[str]:
     From that it constructs a rustworkx directed graph.
+    Resources are represented as nodes and links as edges.
 
     Args:
-        resptr_links: A list of objects representing a direct link (resptr)
-                      between a starting resource and a target resource
-        xml_links: A list of objects representing one or more links from a single starting resource
-                   to a set of target resources, where all target resources are linked to
-                   from a single text value on the starting resource.
-        all_resource_ids: IDs of all the resources in the graph
+        resptr_links: objects representing a direct link between a starting resource and a target resource
+        xml_links: objects representing one or more links from a single text value of a single starting resource
+                   to a set of target resources
+        all_resource_ids: IDs of all resources in the graph
 
     Returns:
-        The rustworkx graph and a dictionary that contains the index number of the nodes with the original resource id
+        - The rustworkx graph.
+        - A dictionary that maps the rustworkx index number of the nodes to the original resource ID from the XML file.
+        - A list with all the edges in the graph.
     """
     graph: rx.PyDiGraph[Any, Any] = rx.PyDiGraph()  # pylint: disable=no-member
     nodes = [(id_, None, None) for id_ in all_resource_ids]
     node_indices = list(graph.add_nodes_from(nodes))
-    node_id_lookup = dict(zip(all_resource_ids, node_indices))
-    node_index_lookup = dict(zip(node_indices, all_resource_ids))
+    id_to_rustworkx_index = dict(zip(all_resource_ids, node_indices))
+    rustworkx_index_to_id = dict(zip(node_indices, all_resource_ids))
     edges: list[tuple[int, int, ResptrLink | XMLLink]] = [
-        (node_id_lookup[x.source_id], node_id_lookup[x.target_id], x) for x in resptr_links
+        (id_to_rustworkx_index[x.source_id], id_to_rustworkx_index[x.target_id], x) for x in resptr_links
     ]
     for xml in xml_links:
-        edges.extend([(node_id_lookup[xml.source_id], node_id_lookup[x], xml) for x in xml.target_ids])
+        edges.extend([(id_to_rustworkx_index[xml.source_id], id_to_rustworkx_index[x], xml) for x in xml.target_ids])
     graph.add_edges_from(edges)
-    return graph, node_index_lookup, edges, set(node_indices)
+    return graph, rustworkx_index_to_id, edges
 
 
 def _remove_leaf_nodes(
     graph: rx.PyDiGraph[Any, Any],
-    node_index_lookup: dict[int, str],
+    rustworkx_index_to_id: dict[int, str],
     node_indices: set[int],
 ) -> tuple[list[str], set[int]]:
     """
     Leaf nodes are nodes that do not have any outgoing links.
     This means that they have no dependencies and are ok to upload.
-    This function removes them from the graph and the set with remaining nodes in the graph.
+    This function removes them from the graph.
```

Comment:
It isn't a side effect any more, rather a copy of the set is returned. This is easier to understand, plus it doesn't impact performance. Since the return value is documented, I don't think I have to write it here, too.